### ISSUE TITLE
GlobalNet-Configured Exported Headless Service Pod's FQDN Not DNS-Resolvable to Itself

### ISSUE THREAD

**Issue by asanderson** [submariner-io/lighthouse#656]
**What happened**:
GlobalNet-Configured exported headless service pod's FQDN was DNS-resolvable to other joined clusters but NOT resolvable to itself.

**What you expected to happen**:
From within headless service pod, I expected to be able to curl the local cluster's service via `<pod>.<cluster>.<service>.<namespace>.clusterset.local` FQDN.

**How to reproduce it (as minimally and precisely as possible)**:
I deployed a MongoDB replica set using the [bitnami/mongodb helm charts](https://github.com/bitnami/charts/tree/master/bitnami/mongodb) across three Submariner clusters which is too long to describe here; however, a much simpler test could be performed using [Submariner User Guide's Test StatefulSet and Headless Service](https://submariner.io/operations/usage/#test-statefulset-and-headless-service) setup steps and then just curl the exported headless service using the FQDN (e.g. **_curl web-0.cluster3.nginx-ss.nginx-test.svc.clusterset.local_**) from within the pod itself.

**Anything else we need to know?**:
My work-around was to add the `<pod>.<cluster>.<service>.<namespace>.clusterset.local` FQDN to the pod's **_/etc/hosts_** via [hostAliases](https://kubernetes.io/docs/tasks/network/customize-hosts-file-for-pods/).
**Environment**: 
- Diagnose information (use `subctl diagnose all`):  unable to provide since clusters are on private air-gapped (no Internet) bare metal servers
- Gather information (use `subctl gather`):  unable to provide since clusters are on private air-gapped (no Internet) bare metal servers
- Cloud provider or hardware configuration:  private air-gapped (no Internet) bare metal servers (CentOS 7)
- Install tools:  RKE
- Others:  Submariner v0.10.0, Kubernetes v1.19.8

**Comment by vthapar** [comment]
I was able to reproduce locally. Turns out, issue is not that we are unable to resolve the IP, but that we return the Globalnet IP of the local service. Since we don't add any routing rules for globalnet in the local cluster [supposed to use local CIDRs] it doesn't work.

```
 # ping nginx-headless-8588bd5696-l5jl6.cluster3.nginx-headless.default.svc.clusterset.local
PING nginx-headless-8588bd5696-l5jl6.cluster3.nginx-headless.default.svc.clusterset.local (242.254.3.252): 56 data bytes
64 bytes from 242.254.3.252: seq=0 ttl=62 time=0.181 ms
64 bytes from 242.254.3.252: seq=1 ttl=62 time=0.172 ms
^C
--- nginx-headless-8588bd5696-l5jl6.cluster3.nginx-headless.default.svc.clusterset.local ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.172/0.176/0.181 ms
/ # ping nginx-headless-8588bd5696-9vcfr.cluster2.nginx-headless.default.svc.clusterset.local
PING nginx-headless-8588bd5696-9vcfr.cluster2.nginx-headless.default.svc.clusterset.local (242.254.2.252): 56 data bytes
^C
--- nginx-headless-8588bd5696-9vcfr.cluster2.nginx-headless.default.svc.clusterset.local ping statistics ---
7 packets transmitted, 0 packets received, 100% packet loss
```
In the example above I'm running shell in the pod of headless service in `cluster2`. Curl an dping to pod in cluster3 work, but not to self in cluster2. Also, you can notice the IP is resoved to Globalnet IP.

@asanderson Can you confirm if the alias you added was the local cluster IP of the pod and not the globalnet IP?

**Comment by asanderson** [comment]
@vthapar Actually, my **_hostAliases_** setting is for `ip: 127.0.0.1` which overrides the **_/etc/host 127.0.0.1 localhost_** entry in the pod itself.

**Comment by vthapar** [comment]
@asanderson Looking at the code, it should be unable to resolve any other pod in the local cluster too. Can you confirm it is so? And just to ensure it is same issue, can you confirm that for you also curl [or try ping] resolves query to Globalnet IP and curl/ping to that IP then fails?

**Comment by asanderson** [comment]
@vthapar Just to clarify, it **_DOES NOT_** resolve other pods in the same cluster using the globalnet fqdn format `<pod>.<cluster>.<service>.<namespace>.svc.clusterset.local`, but it **_DOES_** resolve other pods in a joined cluster using the globalnet FQDN as well as the same cluster using `<service>.<namespace>.svc.clusterset.local` **_AND_** `<pod>.<service>.<namespace>.svc.cluster.local` **_BUT_** the latter is put in the **_/etc/hosts_** file by the [bitnami/mongodb helm chart](https://github.com/bitnami/charts/tree/master/bitnami/mongodb).

Unfortunately, the [bitnami/bitnami-docker-mongodb](https://github.com/bitnami/bitnami-docker-mongodb) image does not include ping nor dig, so I can't confirm ping but all of the above was tested using curl.

Hope that helps.

**Comment by vthapar** [comment]
@asanderson Thanks, this confirms the root cause and am working on possible fixes. I'll push a PR once I have something good enough.

**Comment by dfarrell07** [comment]
@vthapar is this still open? I see the release notes label, which seems to imply something was merged.

**Comment by vthapar** [comment]
@dfarrell07 Yeah, it is still open, was low priority for 0.12.0. Not sure why the release-note label, maybe if we also capture known open issues in release notes?

**Comment by sridhargaddam** [comment]
@vthapar do you think we should target this bug for next release?

**Comment by mikenorgate** [comment]
Is there any updates on this as I'm blocked by this issue. I use a headless service to identify the addresses of all instances of a service that are running in all clusters, due to the lookup only returning globalnet addresses I'm only able to communicate with instances in other clusters
